[5주차|과제1] 레스토랑 조회 구현하기 코드 리뷰 부탁드립니다.

jiunkoo opened the pull request:
PR 열어놓고 과제 진행하도록 하겠습니다.

1. 기본적인 App 개발
    카테고리 구현(완료)
    지역 구현(완료)
    지역에 버튼 클릭 이벤트(완료)
    카테고리에 버튼 클릭 이벤트(완료)
    레스토랑 구현(완료)

2. 서버에서 데이터 가져오기
    카테고리 데이터 서버에서 가져오기(완료)
    지역 데이터 서버에서 가져오기(완료)
    레스토랑 데이터 서버에서 가져오기(완료)

3. 테스트 커버리지 100%(완료)

jiunkoo committed:
npm 패키지 설치

jiunkoo committed:
App.test.jsx 파일 생성

jiunkoo committed:
App - 테스트 추가

jiunkoo committed:
App - 테스트 구현

jiunkoo committed:
Restaurants.jsx 파일 생성

jiunkoo committed:
App에서 레스토랑 목록 Restaurants로 분리

jiunkoo committed:
Restaurants.test.jsx 파일 생성

jiunkoo committed:
Restaurants - 레스토랑 목록 테스트 추가

jiunkoo committed:
Restaurants - 레스토랑 목록 테스트 구현

jiunkoo committed:
Restaurants - 레스토랑 목록 테스트 리팩토링

jiunkoo committed:
RestaurantsContainer.jsx, RestaurantsContainer.test.jsx 파일 생성

jiunkoo committed:
__mocks__ 폴더, resct-redux.js 파일 생성

jiunkoo committed:
fixtures 폴더, restaurants.js 파일 생성 - 테스트 데이터 분리

jiunkoo committed:
RestaurantsContainer - 레스토랑 목록 테스트 추가

jiunkoo committed:
RestaurantsContainer - 레스토랑 목록 테스트 구현

jiunkoo committed:
App - RestaurantsContainer 컴포넌트 붙이기

jiunkoo committed:
App - 테스트 리팩토링

jiunkoo committed:
reducer.js, reducer.test.js 파일 생성

jiunkoo committed:
Restaurants -> Categories 변경

jiunkoo commented:
강사님의 조언에 따라 5주차 과제 내용으로 진행하던 코드를 약간 변경하였습니다!

hannut91 commented:
레스토랑 테스트에서 갑자기 카테고리 테스트로 바꼈군요! 잘 진행하고 계세요 ㅎㅎ 그대로 계속 진행해주세요.

jiunkoo committed:
actions - setCategories 테스트 추가

jiunkoo committed:
actions - setCategories 테스트 구현

jiunkoo committed:
reducer - setCategories 테스트 추가

jiunkoo committed:
reducer - setCategories 테스트 구현

jiunkoo committed:
App - loadCategories 테스트 추가

jiunkoo committed:
App - loadCategories 테스트 구현

jiunkoo committed:
services 폴더, api.js 파일 생성

jiunkoo committed:
api - fetchCategories api 구현

jiunkoo committed:
App - fetchCategories api 붙이기

jiunkoo committed:
__mocks__ 폴더, api.js 파일 생성

jiunkoo committed:
App - loadCategories 테스트 리팩토링

jiunkoo committed:
store.js 파일 생성 및 리덕스 설정

jiunkoo committed:
store -redux - thunk 추가

jiunkoo committed:
actions - loadCategories 테스트 추가

jiunkoo committed:
actions - loadCategories 테스트 구현

jiunkoo committed:
App - loadCategories 분리

jiunkoo committed:
App - loadCategories 분리 및 테스트 리팩토링

jiunkoo committed:
Regions.jsx, Regions.test.jsx, RegionsContainer.jsx, RegionsContainer.test.jsx 파일 생성

jiunkoo committed:
regions 테스트 파일 생성

jiunkoo committed:
Regions - 지역 목록 테스트 추가

jiunkoo committed:
Regions - 지역 목록 테스트 구현

jiunkoo committed:
RegionsContainer - 지역 목록 테스트 추가

jiunkoo committed:
RegionsContainer - 지역 목록 테스트 구현

jiunkoo committed:
api - fetchRegions api 구현

jiunkoo committed:
actions - setRegions, loadRegions 테스트 추가

jiunkoo committed:
actions - setRegions, loadRegions 테스트 구현

jiunkoo committed:
reducer - setRegions 테스트 추가

jiunkoo committed:
reducer -setRegions 테스트 구현

jiunkoo committed:
App - RegionsContainer 컴포넌트 붙이기

jiunkoo committed:
App - 테스트 리팩토링

jiunkoo committed:
Categories - 테스트 리팩토링

jiunkoo committed:
eslint 에러 수정

jiunkoo committed:
Button.jsx, Button.test.jsx 파일 생성

jiunkoo committed:
Button - 테스트 추가

jiunkoo committed:
Button - 테스트 구현

wholemann commented on src/services/api.js:
```diff
@@ -0,0 +1,2 @@
+export default function api(){
+}
```
파일 생성 자체는 어떤 의미를 갖진 못해서

commit 단위로 나눌 필요까진 없어요~

jiunkoo replied:
넵넵

wholemann commented on src/Button.test.jsx:
```diff
@@ -0,0 +1,41 @@
+import Button from './Button';
+
+describe('Button', () => {
+  context('상위 컴포넌트에서 전달받은 props 테스트', () => {
```
지금은 context가 등장할 때가 아닌 듯 합니다.

맥락이 나뉘지 않기 때문이죠.

항상 테스트는 간단한 상태로 시작하시기 바랍니다.

jiunkoo replied:
context를 써보고 싶어서 포함시켜 봤는데 맥락이 나뉠 때 쓰는 거네요! 참고하겠습니다!

wholemann replied:
https://www.youtube.com/watch?v=gfTsSBRvdqI&feature=youtu.be

이 영상을 참고해주세요!

wholemann commented on src/Button.jsx:
```diff
@@ -0,0 +1,11 @@
+import React from 'react';
+
+export default function Button({ value: { id, name }, onClick }) {
```
버튼은 어디에 쓰이나요? 범용적으로 쓰일 곳이 있는지 먼저 고민하고 나눠도 늦지 않습니다.

일단 가장 큰 문제는 Button 자체가 아무 의미를 드러내지 못하고 있습니다.

jiunkoo replied:
조금 더 고민해서 분리해보도록 하겠습니다!

wholemann reviewed (commented):
잘하고 계시네요! TDD에 이제 좀 익숙해지신 것 같습니다.

점진적으로 잘 진행해주셨어요.

몇 가지 아쉬운 부분을 comment 남겨드렸으니 잘 참고하셔서 진행해주세요ㅎㅎ

일단 지금은 Button이 등장할 때가 아닙니다ㅎㅎ 최소한의 의미를 갖는 컴포넌트를 고민해보세요.

Button은 디자인 시스템 측면에서 범용적으로 사용될 때 등장합니다.

지금 Region이나 Category가 버튼이 아니게 된다면 어떻게 될까요?

jiunkoo committed:
Categories - Button 컴포넌트 제거, categories 버튼 테스트 추가

jiunkoo committed:
Categories - categories 버튼 테스트 구현

jiunkoo committed:
Regions - regions 버튼 테스트 추가

jiunkoo committed:
Regions - regions 버튼 테스트 구현

jiunkoo committed:
fixture - regions 추가

jiunkoo committed:
actions - region 선택 테스트 추가

jiunkoo committed:
actions - region 선택 테스트 구현

jiunkoo committed:
reducer - region 선택 테스트 추가

jiunkoo committed:
reducer - region 선택 테스트 구현

jiunkoo committed:
RegionsContainer - region 클릭 시 버튼 이벤트 구현

jiunkoo committed:
Regions - 선택한 regions 화면 출력 테스트 추가

jiunkoo committed:
Regions - 선택한 regions 화면 출력 테스트 구현

jiunkoo committed:
RegionsContainer - regions 버튼 클릭 이벤트 구현에 따른 테스트 리팩토링

jiunkoo committed:
fixture - category 추가

jiunkoo committed:
actions - category 선택 테스트 추가

jiunkoo committed:
actions - category 선택 테스트 구현

jiunkoo committed:
reducer - category 선택 테스트 추가

jiunkoo committed:
reducer - category 선택 테스트 구현

jiunkoo committed:
CategoriesContainer - category 클릭 시 버튼 이벤트 구현

jiunkoo committed:
Categories - 선택한 categories 화면 출력 테스트 추가

jiunkoo committed:
Categories - 선택한 categories 화면 출력 테스트 구현

jiunkoo committed:
CategoriesContainer - category 버튼 클릭 이벤트 구현에 따른 테스트 리팩토링

jiunkoo committed:
Coverage 100% 달성을 위한 리팩토링
- App - fixture 변경 및 테스트 리팩토링
- RegionsContainer - dispatch 테스트 추가
- CategoriesContainer - dispatch 테스트 추가
- reducer - initialState와 unexpected action 테스트 추가

jiunkoo committed:
api - fetchRegions, fetchCategories 테스트 추가

jiunkoo committed:
fixture - restaurants 추가

jiunkoo committed:
api - fetchRestaurants 테스트 추가

jiunkoo committed:
api - fetchRestaurants 테스트 구현

jiunkoo committed:
eslint 에러 수정

jiunkoo committed:
Restaurants - 레스토랑 목록 테스트 추가

jiunkoo committed:
Restaurants - 레스토랑 목록 테스트 구현

jiunkoo committed:
RestaurantsContainer - 레스토랑 목록 테스트 추가

jiunkoo committed:
RestaurantsContainer - 레스토랑 목록 테스트 구현

jiunkoo committed:
actions - setRestaurants 테스트 추가

jiunkoo committed:
actions - setRestaurants 테스트 구현

jiunkoo committed:
reducer - setRestaurants 테스트 추가

jiunkoo committed:
reducer - setRestaurants 테스트 구현

jiunkoo committed:
reducer - restaurants 추가에 따른 initialState 리팩토링

jiunkoo committed:
api - __mocks__에 fetchRestaurants 추가

jiunkoo committed:
CategoriesContainer - 카테고리 클릭 시 restaurants dispatch

jiunkoo committed:
CategoriesContainer - 테스트 리팩토링

jiunkoo committed:
App - RestaurantsContainer 컴포넌트 추가

jiunkoo committed:
App - Restaurants 컴포넌트 추가에 따른 리팩토링

jiunkoo committed:
RegionsContainer - loadRestaurants 추가 및 리팩토링

jiunkoo commented:
드디어 통과ㅠㅠ...버튼 클릭 및 restaurants 목록 불러오기는 아샬님 영상을 참고해서 구현했습니다.

wholemann commented on src/Categories.test.jsx:
```diff
@@ -0,0 +1,51 @@
+
+describe('Categories', () => {
+  describe('상위 컴포넌트에서 받은 props 테스트', () => {
+    const onClick = jest.fn();
```
handleClick이겠죠?

wholemann commented on src/Categories.test.jsx:
```diff
@@ -0,0 +1,51 @@
+  describe('상위 컴포넌트에서 받은 props 테스트', () => {
+    const onClick = jest.fn();
+
+    test('value를 버튼 이름으로 출력', () => {
```
it('renders category text', () => ...)

wholemann commented on src/Categories.test.jsx:
```diff
@@ -0,0 +1,51 @@
+      expect(getByText(/한식/)).not.toBeNull();
+    });
+
+    test('버튼 동작 테스트', () => {
```
it('listens click event', () => ...)

wholemann commented on src/Categories.test.jsx:
```diff
@@ -0,0 +1,51 @@
+import { categories, selectedCategory } from '../fixtures/categories';
+
+describe('Categories', () => {
+  describe('상위 컴포넌트에서 받은 props 테스트', () => {
```
사실 이 테스트는

it('renders list of categories and listens click event', () => ...)

이거 하나로 합쳐도 문제 없을 거에요.

wholemann commented on src/Categories.test.jsx:
```diff
@@ -0,0 +1,51 @@
+        />
+      ));
+
+      expect(container).toHaveTextContent('한식(V)');
```
이 부분은 사실 컴포넌트 테스트에서 표현될 부분은 아닙니다.

컴포넌트는 그저 handleClick의 호출로 인한 action으로 redux에서 변경된 상태를 반영할 뿐이죠.

아샬님 강의에서 이러한 테스트를 하지 않는 이유죠ㅎㅎ

jiunkoo replied:
아하...! 말씀 감사합니다.

wholemann commented on src/CategoriesContainer.test.jsx:
```diff
@@ -0,0 +1,32 @@
+import { selectCategory } from './actions';
+
+describe('CategoriesContainer', () => {
+  it('handleClick 테스트', () => {
```
it('listens click event', () => ...)

위와 같이 완결된 서술이 필요합니다.

테스트의 의도를 명확하게 표현해주세요.

wholemann commented on src/Regions.jsx:
```diff
@@ -0,0 +1,16 @@
+        <li key={region.id}>
+          <button type="button" onClick={() => onClick(region.id)}>
+            {region.name}
+            {selectedRegion && region.id === selectedRegion.id ? '(V)' : null }
```
```suggestion
           {selectedRegion ? (
              <>
                {region.id === selectedRegion.id ? '(V)' : null}
              </>
            ) : null}
```

wholemann commented on src/services/api.test.js:
```diff
@@ -0,0 +1,46 @@
+      await mockFetch(regions);
+      const data = await fetchRegions();
+
+      expect(data).toEqual(regions);
```
저는 regions가 직접적으로 어떤 데이터 구조인지 표현되면 읽는 입장에서 더 좋을 것 같네요.

wholemann reviewed (commented):
잘해내셨습니다ㅎㅎ

아샬님 강의를 다시 보시고 쭉 따라서 복습해보시면

어떤 부분을 더 집중해서 테스트해야 하고, 각 테스트 별 관심사가 무엇인지

파악하실 수 있을 겁니다.

일단 연습은 극단적으로 하는 걸 추천드립니다만

실무적으론 항상 균형과 각 테스트의 경계를 생각하며 테스트 하는 게 좋습니다.

추가적인 질문이 있으시면 언제든 DM 주세요!

jiunkoo committed:
[리팩토링] 코드 리뷰에 따라 리팩토링
- 전체적으로 테스트 의도를 명확하게 표현하기 위해 서술 변경
- api에서 fixture 제거하고 테스트 데이터 집어넣기
- category와 regions의 삼항연산자 변경

jiunkoo commented:
시간이 조금 지났지만 코드 리뷰에 따라 리팩토링 했습니다. 확인 후 merge 부탁드립니다!

wholemann reviewed (approved):
잘하셨습니다. 시간이 지났어도 끝까지 마무리 해보신 게 좋네요:)

Merge 하겠습니다😎
